File: php/15-3sum.php
<?php
class Solution 
{
    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    public function threeSum($nums) 
    {
        $len = count($nums);
        sort($nums);
        $k = 0;
        $result = [];
        while ($k < $len) {
            if ($k > 0 && $nums[$k-1] > 0) break;
            $i = $k+1;
            $j = $len-1;
            $target = 0 - $nums[$k];
            if ($k > 0 && $nums[$k] == $nums[$k-1]) {
                $k++;
                continue;
            }
            while ($i < $j) {
                if ($nums[$i]+$nums[$j] == $target) {
                    $result[] = [$nums[$k], $nums[$i], $nums[$j]];
                    while ($i < $j && $nums[$i] == $nums[$i+1]) $i++;
                    while ($j > $i && $nums[$j] == $nums[$j-1]) $j--;
                    $i++;
                    $j--;
                } else if ($nums[$i]+$nums[$j] < $target) {
                    $i++;
                } else {
                    $j--;
                }
            }
            $k++;
        }

        return $result;
    }
}

File: go/15-3sum.go
package main

import (
	"sort"
)

func threeSum(nums []int) [][]int {
	var res [][]int
	sort.Ints(nums)
	numLen := len(nums)

	for i := 0; i < numLen-2; i++ {
		if nums[i] > 0 {
			break
		}
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}
		target := 0 - nums[i]
		left := i + 1
		right := numLen - 1
		for left < right {
			sum := nums[left] + nums[right]
			if sum == target {
				ele := []int{nums[i], nums[left], nums[right]}
				res = append(res, ele)
				left++
				right--
				for left < right && nums[left] == ele[1] {
					left++
				}
				for right > left && nums[right] == ele[2] {
					right--
				}
			} else if sum > target {
				right--
			} else {
				left++
			}
		}
	}
	return res
}
